### ISSUE TITLE
Trained discriminator not working when after netD.eval() or shuffle = False

### ISSUE THREAD

**Issue by 1214710638** [aimerykong/OpenGAN#10]
hello, thanks for open source your code. I notice that in demo_CrossDatasetOpenSet_testing.ipynb, you didn't set netD.eval() before testing, and the dataloader definition sets shuffle=True. I train and test the model follow your given demo. however, when i set netD.eval() or shuffle=True in dataloader, the test result is not good, the discriminator is not working at all. am i missing something or any suggestion for me?

**Comment by aimerykong** [comment]
Happy to help diagnose the issue. Can you send along two curves of AUROC at
epochs: one with netD.eval() and the other without? The shuffle thing
should not be crucial.

Shu

**Comment by 1214710638** [comment]
Thanks for your reply. it took me some times to get the two curves. Here they are:
![企业微信截图_1656917892529](https://user-images.githubusercontent.com/24508284/177098950-072d42a8-66bd-4236-b8ae-1482e47bf48c.png)
![企业微信截图_16569179346353](https://user-images.githubusercontent.com/24508284/177099055-75277619-7495-4423-8131-ba1a545b8a45.png)
As you can see, the discriminator is not working properly after netD.eval()

**Comment by 1214710638** [comment]
![企业微信截图_16569180502202](https://user-images.githubusercontent.com/24508284/177099380-3d3a2193-2bdd-4aa7-90a7-638a8762df41.png)
![企业微信截图_16569180758575](https://user-images.githubusercontent.com/24508284/177099450-6311852d-399d-4f37-894c-afb4d30356d5.png)
Also, there are curves for mean confidence scores, you can see the one with netD.eval() is unseparatable.

**Comment by aimerykong** [comment]
Thanks for the curves! I recall that I observed the same in my experiments,
that's why I didn't use .eval(). I conjecture that this is because the
discriminator is trained in the train mode for the best discriminative
between outliers and closed-set; if turning on the eval mode, the
discriminator can't work well. I can't remember if tried this or not -- set
batch size as 1 and compare between the eval and train modes again. Would
you like to try them out?

Shu

**Comment by aimerykong** [comment]
Thanks for the discussion! This is a fair point and I agree that there can
be information leakage in the testing implementation. That made me think
about setting batch size as 1 when testing with the train mode. I'll try
this later; I'd appreciate it if you can help try this by reusing your
trained models.
Shu

**Comment by hutudebug** [comment]
In my experiments, even when turn on .train() mode in test phase, the performance still largely rely on batchsize, the order of test sequence, etc, which makes this idea hard to follow:(

**Comment by libo-huang** [comment]
> In my experiments, even when turn on .train() mode in test phase, the performance still largely rely on batchsize, the order of test sequence, etc, which makes this idea hard to follow:(

When testing with eval mode I also get random results. That is `netD.eval()` in OpenGAN doesn't work for solving open-set problems. Can anyone help to check if we are wrong somewhere?

**Comment by Kanbe-Kotori** [comment]
Hello!

I observed the same thing in my trained model, i.e., the results of the model are essentially random with eval() of the discriminator, and the results of the model are stable and good without eval()

However, **If eval() and testing batchsize is 1, the results do not appear random, and are even better than when testing batchsize > 1 and eval() not turned on!**

For me, the conclusion is obvious: **just set the testing  batchsize to 1, and it has nothing to do with eval()** !

**Comment by Kanbe-Kotori** [comment]
> ![企业微信截图_16569180502202](https://user-images.githubusercontent.com/24508284/177099380-3d3a2193-2bdd-4aa7-90a7-638a8762df41.png) ![企业微信截图_16569180758575](https://user-images.githubusercontent.com/24508284/177099450-6311852d-399d-4f37-894c-afb4d30356d5.png) Also, there are curves for mean confidence scores, you can see the one with netD.eval() is unseparatable.

My guess is that the distribution of open set samples leaks into the bn layer without eval(). While the parameters of the model did not get updated by back-propagation, the bn statistics still did get updated.
